Pressing ⌘V pastes content twice in text fields on Tableau analytics dashboard

whsieh opened the pull request:
#### 2575438ca004115ca5cf1c3225c27a6312b7027d
<pre>
Pressing ⌘V pastes content twice in text fields on Tableau analytics dashboard
<a href="https://bugs.webkit.org/show_bug.cgi?id=263590">https://bugs.webkit.org/show_bug.cgi?id=263590</a>
rdar://105750465

Reviewed by Ryosuke Niwa.

In Tableau&apos;s analytics tool, pressing ⌘V to paste in any focused editable areas pastes content
twice after showing a Paste menu item, if the user clicks &quot;Paste&quot; on this item. This is because
Tableau&apos;s script does something akin to the following:

```
textField.addEventListener(&quot;keydown&quot;, event =&gt; {
    if (event.key === &quot;v&quot; &amp;&amp; event.metaKey)
        document.execCommand(&quot;Paste&quot;);
});
```

...which triggers a programmatic paste upon `keydown`, without preventing default. This means that
if the programmatic DOM paste is accepted, we&apos;ll end up triggering two paste commands: (1) due to
the `execCommand`, and (2) due to the default behavior of ⌘V.

While this is ostensibly a website bug, it works fine in other browsers (Firefox, Chrome) because
they don&apos;t support DOM paste at all, so we just end up silently failing the programmatic paste
before performing the real paste.

For now, fix this by adding a quirk for Tableau&apos;s analytics page which disables DOM paste access
triggered by key events. Making this a quirk limits risk in the short term, since it&apos;s possible that
there are other web apps and frameworks that already assume (based on user agent/engine checks) that
Safari/WebKit will show DOM paste prompts on key events.

* Source/WebCore/dom/UserGestureIndicator.cpp:
(WebCore::UserGestureToken::UserGestureToken):
(WebCore::UserGestureIndicator::UserGestureIndicator):
* Source/WebCore/dom/UserGestureIndicator.h:

Add a new enum flag to determine whether or not we should allow DOM paste requests under the user
gesture token.

(WebCore::UserGestureToken::create):
(WebCore::UserGestureToken::canRequestDOMPaste const):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::internalKeyEvent):

Pass in `CanRequestDOMPaste::No` if the quirk is enabled.

* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::requestDOMPasteAccess):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsDisableDOMPasteAccessQuirk const):

Add the quirk; check `window.tableauPrep` instead of a domain, to fix other (non-Apple-internal)
Tableau instances which would also encounter this same issue.

* Source/WebCore/page/Quirks.h:

Canonical link: <a href="https://commits.webkit.org/269736@main">https://commits.webkit.org/269736@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/412100200c0cf7bc5b5681a7d3862fc3f7b63df7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23393 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1507 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24516 "Built successfully") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21611 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23664 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3052 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23946 "Built successfully") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23634 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1074 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20261 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26157 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/881 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21153 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27298 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21340 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21414 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/25171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/862 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18599 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/819 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5592 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1274 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1123 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1574,4 +1574,24 @@ bool Quirks::shouldDisableDataURLPaddingValidation() const
+        if (!globalObject)
+            return false;
+        auto tableauPrepProperty = JSC::Identifier::fromString(globalObject->vm(), "tableauPrep"_s);
+        return globalObject->hasProperty(globalObject, tableauPrepProperty);
```
This can execute arbitrary scripts and it's script observable. Why not check the domain name first then check this property to limit the impact this quirk detection has?

whsieh replied:
Good point — the intent was to make this quirk cover all Tableau instances, ~but I can limit it by host name (tableauanalytics.corp.apple.com) to avoid exposing any script-observable behaviors here.~

Thanks for the review!

rniwa replied:
Oh, Tableau can be deployed on third party domains? If that's the case I suppose the current quirk is okay.

whsieh replied:
Indeed! Okay — I think I'll preserve the current behavior then (without limiting to domain as well). At least, this quirk check _only_ happens when we're about to dispatch DOM key events anyways, so it shouldn't introduce any weird new opportunities for arbitrary script execution.

whsieh committed:
Pressing ⌘V pastes content twice in text fields on Tableau analytics dashboard
https://bugs.webkit.org/show_bug.cgi?id=263590
rdar://105750465

Reviewed by Ryosuke Niwa.

In Tableau's analytics tool, pressing ⌘V to paste in any focused editable areas pastes content
twice after showing a Paste menu item, if the user clicks "Paste" on this item. This is because
Tableau's script does something akin to the following:

```
textField.addEventListener("keydown", event => {
    if (event.key === "v" && event.metaKey)
        document.execCommand("Paste");
});
```

...which triggers a programmatic paste upon `keydown`, without preventing default. This means that
if the programmatic DOM paste is accepted, we'll end up triggering two paste commands: (1) due to
the `execCommand`, and (2) due to the default behavior of ⌘V.

While this is ostensibly a website bug, it works fine in other browsers (Firefox, Chrome) because
they don't support DOM paste at all, so we just end up silently failing the programmatic paste
before performing the real paste.

For now, fix this by adding a quirk for Tableau's analytics page which disables DOM paste access
triggered by key events. Making this a quirk limits risk in the short term, since it's possible that
there are other web apps and frameworks that already assume (based on user agent/engine checks) that
Safari/WebKit will show DOM paste prompts on key events.

* Source/WebCore/dom/UserGestureIndicator.cpp:
(WebCore::UserGestureToken::UserGestureToken):
(WebCore::UserGestureIndicator::UserGestureIndicator):
* Source/WebCore/dom/UserGestureIndicator.h:

Add a new enum flag to determine whether or not we should allow DOM paste requests under the user
gesture token.

(WebCore::UserGestureToken::create):
(WebCore::UserGestureToken::canRequestDOMPaste const):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::internalKeyEvent):

Pass in `CanRequestDOMPaste::No` if the quirk is enabled.

* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::requestDOMPasteAccess):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsDisableDOMPasteAccessQuirk const):

Add the quirk; check `window.tableauPrep` instead of a domain, to fix other (non-Apple-internal)
Tableau instances which would also encounter this same issue.

* Source/WebCore/page/Quirks.h:

Canonical link: https://commits.webkit.org/269736@main
